neural_cherche/retrieve/bm25.py

```diff
@@ -0,0 +1,203 @@
+
+from sklearn.feature_extraction.text import CountVectorizer
+from scipy.sparse import csc_matrix, csr_matrix
+import numpy as np
+from .. import utils
+__all__ = ["BM25"]
+
+class BM25:
+    """BM25 .
+
+    Parameters
+    ----------
+    key
+        Field identifier of each document.
+    on
+        Fields to use to match the query to the documents.
+    documents
+        Documents in TFIdf retriever are static. The retriever must be reseted to index new
+        documents.
+    k
+        Number of documents to retrieve. Default is `None`, i.e all documents that match the query
+        will be retrieved.
+    CountVectorizer
+        CountVectorizer class of Sklearn to create a custom CountVectorizer counter.
+
+    Examples
+    --------
+    >>> from neural_cherche import retrieve
+    >>> from pprint import pprint
+
+    >>> documents = [
+    ...     {"id": 0, "document": "Food"},
+    ...     {"id": 1, "document": "Sports"},
+    ...     {"id": 2, "document": "Cinema"},
+    ... ]
+
+    >>> queries = ["Food", "Sports", "Cinema"]
+
+    >>> retriever = retrieve.bm25(
+    ...     key="id",
+    ...     on=["document"],
+    ... )
+
+    >>> documents_embeddings = retriever.encode_documents(
+    ...     documents=documents
+    ... )
+
+    >>> retriever = retriever.add(
+    ...     documents_embeddings=documents_embeddings,
+    ... )
+
+    >>> queries_embeddings = retriever.encode_queries(
+    ...     queries=queries
+    ... )
+
+    >>> scores = retriever(
+    ...     queries_embeddings=queries_embeddings,
+    ...     k=4
+    ... )
+
+    >>> pprint(scores)
+    [[{'id': 0, 'similarity': 1.0}],
+     [{'id': 1, 'similarity': 0.9999999999999999}],
+     [{'id': 2, 'similarity': 0.9999999999999999}]]
+
+    References
+    ----------
+    1. [sklearn.feature_extraction.text.CountVectorizer](https://scikit-learn.org/stable/modules/generated/sklearn.feature_extraction.text.CountVectorizer.html)
+    2. [Python: tf-idf-cosine: to find document similarity](https://stackoverflow.com/questions/12118720/python-tf-idf-cosine-to-find-document-similarity)
+
+    """
+    def __init__(
+        self,
+        key: str,
+        on: list[str],
+        b=0.75, 
```

Comment:
You can add type hinting here, b: float = 0.75, k: float = 1.5, count_vectorizer: CountVectorizer